crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,823 @@
+///   sorted. The string data attached to this variant is the source
+///   code of the sorted `__all__`, that can be inserted into the source
+///   code as a `range_replacement` autofix.
+#[derive(Debug)]
+enum SortedDunderAll {
+    AlreadySorted,
+    Sorted(String),
+}
+
+/// Collect data on each line of `__all__`.
+/// Return `None` if `__all__` appears to be invalid,
+/// or if it's an edge case we don't care about.
+///
+/// Why do we need to do this using the raw tokens,
+/// when we already have the AST? The AST strips out
+/// crucial information that we need to track here, such as:
+/// - The value of comments
+/// - The amount of whitespace between the end of a line
+///   and an inline comment
+/// - Whether or not the final item in the tuple/list has a
+///   trailing comma
+///
+/// All of this information is necessary to have at a later
+/// stage if we're to sort items without doing unnecessary
+/// brutality to the comments and pre-existing style choices
+/// in the original source code.
+fn collect_dunder_all_lines(
+    range: TextRange,
+    locator: &Locator,
+) -> Option<(Vec<DunderAllLine>, bool)> {
+    // These first three variables are used for keeping track of state
+    // regarding the entirety of the `__all__` definition...
+    let mut parentheses_open = false;
+    let mut ends_with_trailing_comma = false;
+    let mut lines = vec![];
+    // ... all state regarding a single line of an `__all__` definition
+    // is encapsulated in this variable
+    let mut line_state = LineState::default();
+
+    // `lex_starts_at()` gives us absolute ranges rather than relative ranges,
+    // but (surprisingly) we still need to pass in the slice of code we want it to lex,
+    // rather than the whole source file:
+    for pair in lexer::lex_starts_at(locator.slice(range), Mode::Expression, range.start()) {
+        let (tok, subrange) = pair.ok()?;
+        match tok {
+            // If exactly one `Lpar` or `Lsqb` is encountered, that's fine
+            // -- a valid __all__ definition has to be a list or tuple,
+            // and most (though not all) lists/tuples start with either a `(` or a `[`.
+            //
+            // Any more than one `(` or `[` in an `__all__` definition, however,
+            // indicates that we've got something here that's just too complex
+            // for us to handle. Maybe a string element in `__all__` is parenthesized;
+            // maybe the `__all__` definition is in fact invalid syntax;
+            // maybe there's some other thing going on that we haven't anticipated.
+            //
+            // Whatever the case -- if we encounter more than one `(` or `[`,
+            // we evidently don't know what to do here. So just return `None` to
+            // signal failure.
+            Tok::Lpar | Tok::Lsqb => {
+                if parentheses_open {
+                    return None;
+                }
+                parentheses_open = true;
+            }
```

Comment:
> How common are unparenthesized tuples for `__all__`? I thought `__all__` is recommended to be a list. Maybe it's not worth supporting fixing tuples?

Tuple `__all__`s are very common; I'd definitely rather support them here. I don't think there's a recommendation w.r.t. whether they should be tuples or lists -- I'm not actually sure which is more common. Conceptually, I sort-of prefer using a tuple for `__all__`, personally. But I think that's just a vague instinct towards preferring immutable types for global constants, rather than something that's founded on any strong reasoning :)

Using an _unparenthesized_ tuple for `__all__` is less common, but not unheard of:

- https://github.com/python/cpython/blob/f8a79109d0c4f408d34d51861cc0a7c447f46d70/Lib/asyncio/base_events.py#L52
- https://github.com/python/cpython/blob/f8a79109d0c4f408d34d51861cc0a7c447f46d70/Lib/asyncio/coroutines.py#L1
- https://github.com/python/cpython/blob/f8a79109d0c4f408d34d51861cc0a7c447f46d70/Lib/asyncio/windows_utils.py#L17

Reply:
What happened to 

> There should be one-- and preferably only one --obvious way to do it.

Reply:
> What happened to
> 
> > There should be one-- and preferably only one --obvious way to do it.

many people are saying this

Reply:
Well, I'd quite like to use that `is_tuple_parenthesized` function you've got there in the formatter... what's the best course of action there? Copy-and-paste it, or move it to `ruff_python_trivia`?

Reply:
You could add it to `ruff_python_ast` and either implement it as a free-standing function or as a method on `ExprTuple`.